cachito/workers/scm.py

```diff
@@ -62,15 +63,50 @@ def _reset_git_head(self, repo):
                 f'of "{self.ref}" is valid.'
             )
 
+    def _verify_archive(self):
+        """
+        Verify the archive containing the git repository.
+
+        :raises CachitoError: if 'git fsck' fails for the extracted sources
+        """
+        log.debug("Verifying the archive at %s", self.sources_dir.archive_path)
+        if not os.path.exists(self.sources_dir.archive_path) or not tarfile.is_tarfile(
+            self.sources_dir.archive_path
+        ):
+            raise CachitoError(f"No valid tar archive at {self.sources_dir.archive_path}")
+
+        with tempfile.TemporaryDirectory(prefix="cachito-") as temp_dir:
+            with tarfile.open(self.sources_dir.archive_path, mode="r:gz") as tar:
+                tar.extractall(temp_dir)
+
+            cmd = ["git", "fsck"]
+            repo_path = os.path.join(temp_dir, "app")
+            try:
+                subprocess.run(cmd, cwd=repo_path, check=True)
+            except subprocess.CalledProcessError as exc:
+                log.error(
+                    "Cachito found an error when verifying the generated archive at '%s': %s",
+                    self.sources_dir.archive_path,
+                    exc,
+                )
+                raise CachitoError(f"Invalid tar archive at {self.sources_dir.archive_path!s}")
+
     def _create_archive(self, from_dir):
         """
-        Create an archive from a specified directory.
+        Create a verified archive from a specified directory.
 
         :param str from_dir: path to a directory from where to create the archive.
+        :raises CachitoError: if the archive verification fails
         """
         log.debug("Creating the archive at %s", self.sources_dir.archive_path)
         with tarfile.open(self.sources_dir.archive_path, mode="w:gz") as bundle_archive:
             bundle_archive.add(from_dir, "app")
+        try:
+            self._verify_archive()
+        except CachitoError:
+            log.debug("Removing invalid archive at %s", self.sources_dir.archive_path)
+            os.unlink(self.sources_dir.archive_path)
+            raise
```

Comment:
Should this raise CachitoError or fall back on re-cloning the repo?

Reply:
It should re-raise the CachitoError at some point.

I thought about retrying the operation, but then I thought these errors would just be forgotten in the logs in case they start happening often and they could hide issues with the file system or the OCP cluster.

I'd be happy to change it to log a warning and retry the operation N times (once should be enough) before raising here.

Thoughts?

Reply:
IMO if the data is corrupted immediately after creation, something is very wrong and I would rather have the request fail. On the off-chance that a retry would work, it would just hide a potentially serious issue as Athos said.

Reply:
Hmmm. I'm fine with raising this issue without a re-try. If we are checking the verification here, and are planning to clean up the existing archives explicitly, why do we need to have this check for existing archives? (out of curiosity)

Reply:
We will run a script in our production environment to clean the corrupted files present at the moment. After this is merged, no more corrupted caches should be created. Even if it happens (let's say a worker is killed while writing a file), next time the same reference is requested, the archive will be re-created. If a different reference is requested, the corrupted file will not be used.